libs/tx-builder/src/utils/deepDecoding.ts

```diff
@@ -0,0 +1,275 @@
+
+export const deepDecodeProposalActions = async ({
+  chainId,
+  actionData,
+  rpcs = HAUS_RPC,
+  explorerKeys = ABI_EXPLORER_KEYS,
+}: {
+  chainId: ValidNetwork;
+  actionData: string;
+  rpcs?: Keychain;
+  explorerKeys?: Keychain;
+}): Promise<(DeepDecodedAction | ActionError)[]> => {
+  const abiLoader = new MultiABILoader([
+    new EtherscanABILoader({
+      chainId,
+      rpcs,
+      explorerKeys,
+    }),
+    new SourcifyABILoader(),
+  ]);
+
+  const options = {
+    chainId,
+    rpcs,
+    explorerKeys,
+    loader: abiLoader,
+  };
+
+  return decodeMultiCall(options, actionData as `0x${string}`);
+};
+
+const decodeMultiCall = async (
+  options: Options,
+  data: `0x${string}`
+): Promise<DeepDecodedMultiTX> => {
+  const proposalActions: MetaTransaction[] = decodeMulti(data);
+
+  const decodedProposalActions = await Promise.all(
+    proposalActions.map(async (action) => {
+      try {
+        return await decodeAction(options, action);
+      } catch (e) {
+        return {
+          error: true,
+          message: (e as Error).message,
+          data,
+        };
+      }
+    })
+  );
+
+  return decodedProposalActions;
+};
+
+type FunctionDataWithInputsReturnType = {
+  functionName: string;
+  inputs: {
+    name: string;
+    type: string;
+    value: string;
+  }[];
+};
+
+const decodeValue = (value: unknown): string => {
+  if (typeof value === 'string') {
+    if (value.startsWith('0x')) {
+      return fromHex(value as `0x${string}`, 'bigint').toString();
+    }
+    return value;
+  }
+  if (typeof value === 'number' || typeof value === 'bigint') {
+    return BigInt(value).toString();
+  }
+  return '0';
+};
+
+const decodeFunctionDataWithInputs = (options: {
+  abi: any[];
+  data: `0x${string}`;
+}): FunctionDataWithInputsReturnType => {
+  const result = decodeFunctionData(options);
+
+  const functionDetails = getAbiItem({
+    abi: options.abi,
+    name: result.functionName,
+  });
+
+  const inputs = functionDetails['inputs'] || [];
+
+  const inputsWithValues = (inputs as any[]).map((input, index) => ({
+    name: input.name,
+    type: input.type,
+    value: result.args?.[index]?.toString() || '0x',
+  }));
+
+  return {
+    functionName: result.functionName,
+    inputs: inputsWithValues,
+  };
+};
+
+const decodeAction = async (
+  options: Options,
+  action: MetaTransaction
+): Promise<DeepDecodedAction | ActionError> => {
+  const { data, to, value, operation } = action;
+  const { chainId, rpcs, loader } = options;
+  if (
+    !data ||
+    data === '0x' ||
+    !data.startsWith('0x') ||
+    data === ENCODED_0X0_DATA ||
+    (await getCode({ chainId, contractAddress: action.to, rpcs })) === '0x'
+  ) {
+    return {
+      to: to,
+      operation: operation || OperationType.Call,
+      name: `${HAUS_NETWORK_DATA[chainId]?.symbol} Transfer`,
+      value: decodeValue(value),
+      params: [],
+      decodedActions: [],
+    };
+  }
+
+  const { abi } = await whatsabi.autoload(to, {
```

Comment:
I've never used whatsabi. looks powerful.

After this we might want to explore replacing all our abi fetching with this.

Reply:
Thanks to @dekanbro for the library!